Updated refarch section to sync with supervisor domain spec (#34)
riscv-non-isa/riscv-ap-tee: specification/refarch.adoc

[[refarch]]
== Reference Architecture Details

We describe the capabilities of the platform to support memory isolation
requirements for confidentiality of workloads in TVMs. We then describe
the properties of the TSM, its instantiation, isolation and operational model
for the TVM life cycle. The description in this section refers to the reference
architecture in Figure 1.

> 

Reviewer: This change seems to be just removal of white space. Would be good to put such editorial commits in a separate single commit. Also having one change per commit helps review commit by commit.

=== CoVE Memory Isolation

Memory isolation for TVMs is orchestrated by the TSM-driver and the TSM in two
phases: the conversion of memory to confidential memory and the assignment of
confidential memory (alongwith the enforcement of properties on use) to TVMs.
To enforce isolation across Host and Confidential supervisor domains, CoVE

> requires page-based isolation of physical memory. There are two deployment

Author: ```suggestion
requires isolation of physical memory (that supports paging when enabled). There are two deployment
```

> models possible here. CoVE ABI is equally applicable for both modes - this

Reviewer: what is implied by page-based?

Author: meant to say physical memory isolation that can support page size isolation when paging is enabled - updated text.

specification focusses on the second deployment model (b) where a peer

> supervisor domain is used to host confidential workloads.

Author: I think its good to call out that the spec focusses on use of supervisor domains for CC

a. When the TSM is the only root HS mode component on the platform, the G-stage
page table can be used to enforce isolation between confidential TVMs and
ordinary VMs. In this model the host VMM must execute in the de-privileged VS
mode and the TSM must provide nested virtualization of the H-extension controls.
This model may be suitable for client/embedded systems.

b. The TSM operates in S/HS mode as a peer supervisor domain manager to the
hosting supervisor domain which operates in S/HS mode as well. This model uses

> the MTT along with Sv and G-stage PT for confidential TVM isolation. The MTT

Author: ```suggestion
the MTT along with G-stage PT for confidential TVM isolation (where 1st stage PT is 
used by the Guest OS normally). The MTT
```

> is used to assign physical memory to the Confidential supervisor domain called

Reviewer: Sv is not used for TVM isolation. On G-stage.

Author: meant to say Sv may be used by the guest OS as normal (updated text)

*Confidential* memory and memory accessible to the hosting supervisor domain
called *Non-Confidential*. MTT allows dynamic programming of the per-domain

> access permissions. A TVM needs to access both types of memory:

Author: ```suggestion
access permissions. A TVM and/or TSM needs to access both types of memory:
```

> 

Reviewer: TVM and/or TSM

Author: ack

* Confidential memory - used for TVM code, data

> * Non-Confidential memory - used for communication between TVM and the untrusted
> host entities

Author: ```suggestion
* Non-Confidential memory - used for communication between TVM and the non-TCB
host software and/or non-TCB IO devices.
```

> 

Author: updated.

The TSM COVH ABI provides interfaces to the OS/VMM to convert /
donate memory to confidential supervisor domain and reclaim memory back from the
confidential supervisor domain. TVM memory is by default assigned from
Confidential memory regions. TVM may be assigned shared memory regions as well.

> The TSM enforces at which stage of the TVM memory may be assigned to the TVM.

Reviewer: This sentence does not read right.

Author: fixed ptal

> The TSM COVH ABI provides interfaces to the OS/VMM to convert /
> donate memory to confidential supervisor domain and reclaim memory back from the
> confidential supervisor domain. TVM memory is by default assigned from
> Confidential memory regions. TVM may be assigned shared memory regions as well.
> The TSM enforces at which stage of the TVM memory may be assigned to the TVM.

Author: ```suggestion
The TSM COVH ABI provides interfaces to the OS/VMM to convert / donate 
memory from the hosting supervisor domain to the confidential supervisor domain. 
Similarly, a separate ABI intrinsic is used to reclaim memory back from the 
confidential supervisor domain to the hosting supervisor domain. Once physical 
memory is converted to confidential - it is accessible only to the confidential 
supervisor domain. By default, TVM memory is assigned by the TSM (which 
operates in the confidential supervisor domain context) from confidential 
physical memory regions. Explicitly, a TVM may be assigned shared (non-confidential) 
memory regions as well. The TSM manages the type and accessibility of all memory 
assigned to the TVM, to mitigate attacks from non-TCB software.
```

only for Non-confidential Memory

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Confidential Memory conversion"]
image::img_7.png[]

[NOTE]
====
To also provide physical memory protection, memory accessible to the
confidential supervisor domain via MTT may also be associated with a unique
memory encryption key. Additionally each TVM may also be associated with a
unique memory encryption key. These additional protection aspects are platform
and implementation dependent.
====

Confidential and non-confidential memory are both always assigned by the VMM -
the TSM and TSM-driver are expected to manage the isolation for confidential
memory by programming the Memory Tracking Table (MTT). The desired security
properties of memory tracking are discussed below. The TSM manages
finer-granular (page-based) allocation from Confidential memory regions
(enforced by the memory tracking hardware) using the G-stage page table.

Four aspects of memory isolation are impacted due to this dynamic configurable
property of the MTT:

==== Address Translation/Page Walk
The figure 2 below describes a reference model for memory tracking lookup where

> the physical address derived from the first and guest stage nested page walk is

Author: ```suggestion
the physical address derived from the two-stage address translation and protection mechanism is
```

> looked up via the MTT configured for the active supervisor domain to get the

Reviewer: first and guest stage nested page walk -> two-stage address translation and protection

Author: ack

> access permissions for the physical address. This lookup should be performed
> for the page sizes per the paging modes supported by the hart.

Author: ```suggestion
access permissions for the physical address. This lookup should be performed 
for each implicit and explicit memory access and per the paging sizes/modes 
supported by the hart.
```

Reviewer: "and per paging sizes/modes..." can be removed.

> 

Reviewer: Stonger to say - for each implicit and explicit memory access.

Author: ack

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Memory Tracking for Supervisor Domains"]
image::img_8.png[]

==== Management of isolation for Confidential Physical Memory

The SW TCB (TSM) manages the assignment of physical memory to the Confidential

> supervisor domain, while the HW TCB (hart MMU including Sv, MTT Extensions)

Author: ```suggestion
supervisor domain, while the HW TCB (hart MMU including virtual memory system, MTT Extensions)
```

Reviewer: does MMU already imply virtual memory system. Could it just be "( using page-based virtual memory system and MTT Extensions)"

> enforces the access-control for confidential memory against other supervisor

Reviewer: Sv is not a good architectural term. virtual memory system?

Author: ack

domains. The region sizes at which the memory tracking enforces isolation may
be multiples of the architectural page sizes supported by the hart MMU.
Similarly, the IOMMU via the IOMTT is expected to support a similar memory

> tracking lookup to enable direct IOMMU and IOMMU-managed device access into

Reviewer: IOMMU is not involved with IOMTT. So IOMMU via IOMTT is wrong. It should state IOMMU period.

Reviewer: Is direct acess to IOMMU from TVM a goal? That seems wrong.

Author: first one - fixed. second one - grammar issue - fixed.

> Similarly, the IOMMU via the IOMTT is expected to support a similar memory
> tracking lookup to enable direct IOMMU and IOMMU-managed device access into

Author: ```suggestion
Similarly, the IOMMU is expected to support a similar memory tracking lookup 
to enable a device/function trusted by the TVM to directly access 
```

Reviewer: The IOMMU is not involved with IOMTT or similar memory tracking. 
A better text is 
"Similarly, MTT enforcement for memory accesses originated by I/O devices and IOMMUs"

TVM confidential memory regions. For the CoVE reference architecture this TCB

> consists of the HW (MMU, IOMMU, Memory Controller) and the SW/FW elements -

Author: ```suggestion
consists of the HW (e.g. MMU, IOMMU, Memory Controller) and the SW/FW elements -
```

> TSM-driver and the TSM. The TSM-driver is responsible for enforcing isolation of

Reviewer: Is that definitive list of HW - MMU, IOMMU, Mmeory controller?

Author: added e.g.

confidential memory regions (consisting of multiple pages via MTT) and the TSM
is responsible for enforcing isolation of confidential memory pages among TVMs
(via G-stage translation) - pages assigned to the TVM may be exclusively
accessible to the condidential supervisor domain or may be shared with the
hosting supervisor domain (e.g. to allow for paravirtualized IO access).

[NOTE]
====
The TSM may manage additional attributes on TVM-assigned pages such as:
TVM-owner, Page-sub-type, TLB versioning information, Locking semaphore and
additional metadata etc. This extended memory tracking information managed by the
TSM is referred to as the Extended Memory Tracking Table (EMTT).
====

==== Handling Implicit & Explicit Memory Accesses

> For TVM accesses for instruction fetch and page walks (virtual address
> translation), isolated/confidential memory is required to enforce the
> following security properties:

Author: ```suggestion
For TVM accesses for instruction fetch and page walks, isolated/confidential 
memory is required to enforce the following security properties:
```

Reviewer: Why "must be enforced by hart" and not "enforce by TSM by marking the G-stage page table entries that map non-confdential memory as not executable".

* TEE Instruction fetch - security property: TVM/TSM must not fetch code

> from untrusted/shared memory (this property must be enforced by the hart e.g. MMU)

Reviewer: The text in braces is confusing as it implies there is some special enforcement for this.

Author: removed.

> * TEE Instruction fetch - security property: TVM/TSM must not fetch code
> from untrusted/shared memory (this property must be enforced by the hart e.g. MMU)

Author: ```suggestion
* TEE Instruction fetch - security property: TVM/TSM must not fetch code
from untrusted/shared memory - enforced by the hart e.g. MMU
```

* TEE Paging structure walk - security property: TVM/TSM must not locate
page tables in untrusted shared memory (for TVM, G-stage walks enforced

> to be in confidential memory by the TSM)

Reviewer: The text in braces is redundant following "TVM/TSM must not locate page tables in untrusted memory"

Author: removed text in brackets.

> * TEE Paging structure walk - security property: TVM/TSM must not locate
> page tables in untrusted shared memory (for TVM, G-stage walks enforced
> to be in confidential memory by the TSM)

Author: ```suggestion
* TEE Paging structure walk - security property: TVM and TSM must not locate
page tables in untrusted shared memory.
```

* TEE data fetch - security property: The TVM via the TSM may be allowed to
relax data accesses to non-confidential memory (via MTT) to allow for IO
accesses.

==== Cached translations/TLB management
During confidential memory conversion or reclamation, the TCB must
enforce that stale translations cached in harts are not accessible

> to the untrusted host or another TVM context.

Reviewer: The RISC-V architecture does not define a concept of TLBs architecturally. Architecturally these should be defined as memory-management fences.

Author: fixed

> During confidential memory conversion or reclamation, the TCB must
> enforce that stale translations cached in harts are not accessible
> to the untrusted host or another TVM context.

Author: ```suggestion
During confidential memory conversion or reclamation, the HW TCB 
and SW TCB (TSM) must enforce via memory-management fences 
that stale data is not accessible to the TVM (or the hosting OS/VMM).
```

During confidential memory assignment to a TVM (or during conversion
of confidential memory to shared), the TCB must enforce that stale
translations may not be held to memory yielded by a TVM (and used
by the host for another TVM or VM or the host).
These properties are implemented by the TSM in conjunction with

> the HW (e.g. MTT cache invalidations) via the proposed TEEI.

Author: ```suggestion
the HW (e.g. MTT cache invalidations) via the proposed COVH interface.
```

=== TSM initialization

The CoVE architecture requires a hardware Root-of-trust for supporting
TCB measurement, reporting and storage <<R8>>. The Root-of-trust for
Measurement (RTM) is defined as the TCB component that performs a

> measurement of an entity and protects it for subsequent reporting. The

Author: ```suggestion
measurement of an entity and signs it for subsequent reporting
to a relying party. The
```

> Root-of-trust for Reporting (RTR) is typically a HW RoT that reliably

Reviewer: "protects it" -> should be clarified. Is this refering to protecting the measurement register. A RTM is not responsible for protecting the integrity of the MLE.

Author: fixed it (meant "signs it")

provides authenticity and non-repudiation services for the purposes of
attesting to the origin, integrity and security version of platform TCB
components. Each TCB layer should have associated security version numbers
(SVN) to allow for TCB recovery in the event of security vulnerabilities
discovered in a prior version of the TCB layer.

> During platform initialization, HW elements form the RTM that measure the

Author: ```suggestion
During platform initialization, HW/FW elements form the RTM that measure the
```

TSM-driver. The TSM-driver acts as the RTM for the TSM loaded on the

> platform. The TSM-driver initializes the TSM-memory-region for the TSM -

Reviewer: The "HW elements form the RTM" is a bit overarching?

Author: hw/fw

> this TSM-memory-region must be in TEE-capable memory. The TSM binary may be

Author: ```suggestion
this TSM-memory-region must be in confidential memory. The TSM binary may be
```

provided by the OS/VMM which may independently authenticate the binary
before loading the binary into the TSM-memory-region via the TSM-driver.

> Alternatively, the firmware may pre-load the TSM binary via the TSM-driver.

Reviewer: which firmware?

Author: platform fw

> Alternatively, the firmware may pre-load the TSM binary via the TSM-driver.

Author: ```suggestion
Alternatively, the platform firmware may pre-load the RoT-authenticated TSM binary via the TSM-driver.
```

In both cases, the TSM binary loaded must be measured and may be
authenticated (per cryptographic signature mechanisms) by the TSM-driver
during the loading process, so that the TSM used is reflected in the
attestation rooted in a HW RoT. The authentication process provides
additional control to restrict TSM binaries that can be loaded on the
platform based on policies such as version, vendor etc. In addition to the
measurements, a security version number (SVN) of the TSM should be recorded
by the TSM-driver into the firmware measurement registers accessible only
to the TSM-driver and higher privilege components. The measurements and
versions of the HW RoT, the TSM-driver and the TSM will subsequently be
provided as evidence of a specific TSM being loaded on a specific platform.

During initialization, the TSM-driver will initialize a TSM-data region
within the TSM-memory region. The TSM-data region may hold per-hart TSM
state, memory assignment tracking structures and additional global data for

> TSM management. The TSM-data region is TEE-capable memory that is apriori

Reviewer: TEE-capable or confidential? May want to use a consitent terminology.

Author: fixed to use confidential

> TSM management. The TSM-data region is TEE-capable memory that is apriori

Author: ```suggestion
TSM management. The TSM-data region is confidential memory that is apriori
```

access-control-restricted by the TSM-driver to allow only the TSM to access
this memory. The per-hart TSM state is used to start TSM execution from a
known-good state for security routines invoked by the OS/VMM. The per-hart

> TSM state should be stored in pages that form a TSM Hart Control Structure

Reviewer: pages-> memory?

> TSM state should be stored in pages that form a TSM Hart Control Structure

Author: ```suggestion
TSM state should be stored in confidential memory in a TSM Hart Control Structure
```

(THCS - See <<appendix_a>>) which is initialized as part of the TSM memory

> initialization. The THCS structure definition is part of the TEEI and may

Author: ```suggestion
initialization. The THCS structure definition is part of the COVH ABI and may
```

by the host for another TVM or VM or the host).
These properties are implemented by the TSM in conjunction with
the HW (e.g. MTT cache invalidations) via the proposed COVH interface.

=== TSM initialization

The CoVE architecture requires a hardware Root-of-trust for supporting
TCB measurement, reporting and storage <<R8>>. The Root-of-trust for
Measurement (RTM) is defined as the TCB component that performs a
measurement of an entity and cryptographically signs it as attestation
evidence subsequently reported to a relying party. The
Root-of-trust for Reporting (RTR) is typically a HW RoT that reliably
provides authenticity and non-repudiation services for the purposes of
attesting to the origin, integrity and security version of platform TCB
components. Each TCB layer should have associated security version numbers
(SVN) to allow for TCB recovery in the event of security vulnerabilities
discovered in a prior version of the TCB layer.

During platform initialization, HW/FW elements form the RTM that measure the
TSM-driver. The TSM-driver acts as the RTM for the TSM loaded on the
platform. The TSM-driver initializes the TSM-memory-region for the TSM -

> this TSM-memory-region must be in TEE-capable memory. The TSM binary may be

Author: ```suggestion
this TSM-memory-region must be in confidential memory. The TSM binary may be
```

any v/f register state must be restored by the caller.
* TSM-driver passes TSM/TVM-specified register contents to the OS/VMM to
return status from TEERET (TSM sets a0, a1 registers always - other
registers may be selected by the TVM)
* Clear Confidential mode on hart (via implementation-specific M-mode CSR to
block non-TEE mode accesses to TEE-assigned memory.)
* MRET to resume execution in OS/VMM at mepc set to THCS.hssa.pc
(THCS.hssa.pc adjusted to refer to opcode after the ECALL that triggered
the TEECALL / TEERESUME)

The TSM is stateless across TEECALL invocations, however a security routine
invoked in the TSM via a TEECALL may be interrupted and must be resumed via
a TEERESUME i.e. _the TSM is preemptable but non-reentrant_. These
properties are enforced by the TSM-driver, and other models described above
may be implemented. The TSM does not perform any dynamic resource
management, scheduling, or interrupt handling of its own. The TSM is not
expected
to issue IPIs itself; the TSM must track if appropriate IPIs are issued by the
host OS/VMM to track that the required security checks are performed on each

> physical hart (or virtual hart context) as required by specific TEEI flows.

Author: ```suggestion
physical hart (or virtual hart context) as required by specific COVH/G flows.
```

interrupt causes and handling requirements.

The TSM may not need to program stimecmp on its own, though it may verify
that time is not going back for a TVM. If the TSM needs to start a timer,
it should context switch the stimecmp CSR and replace it with its timeout
value if it's later than the timer it wants to start. The TSM may still
want to be aware of the value programmed into stimecmp to guard against
step attacks on TVMs.

Any NMIs experienced during TSM/TVM execution are always handled by the
TSM-driver and must cause the TEEs to be destroyed (preventing any loss of
confidential info via clearing of machine state). The TSM and therefore all
TVMs are prevented from execution after that point.

=== TSM and TVM Isolation

TSM (and all TVMs) memory is granted by the host OS/VMM but is isolated
(via access-control and/or confidentiality-protection) by the HW and TCB
elements. The TSM, TVM and HW isolation methods used must be evident in the
attestation evidence provided for the TVM since it identifies the hardware
and the TSM-driver.

There are two facets of TVM and TSM memory isolation that are
implementation-specific:

> *a)* *Isolation from host software access* -  The CPU must enforce

Reviewer 2: Is this why you referred to earlier that this spec focuses on host in peer mode deployment ? How about the relaxing this requirement or describe each of these isolation from both deployment models ?

From plumbers discussions, it appeared there are significant interest in host in VS model deployment model as well.

Author: added text for depl. model (a)

> *a)* *Isolation from host software access* -  The CPU must enforce

Author: ```suggestion
*a)* *Isolation from host software access* -  For the deployment model (a), 
the CPU must enforce hardware-based access-control of TSM memory via 
the G-stage page tables to prevent the guest VMM from accessing TSM 
memory. For the deployment model (b). The CPU must also similarly enforce
```

> hardware-based access-control of TSM memory to prevent access from host supervisor domain components
> (VMM and host OS) V=0, HS-mode untrusted code.
> TEE and TVM address spaces are identified by supervisor domain identifiers
> (Smsdid) to maintain the isolation during access and in internal
> address translation caches, e.g. Hart TLB lookup may be extended with the
> SDID in addition to the ASID, VMID for workloads in the Confidential supervisor
> domain. TVM memory isolation must support sparse memory management
> models and architectural page-sizes of 4KB, 64K, 2MB, 1GB (and optionally
> 512GB). The hardware may implement the Smmtt extension, or other approaches may
> be used such as a flat table. The memory tracking table may be enforced at the
> memory controller, or in a page table walker.

Author: ```suggestion
hardware-based access-control of TSM memory to prevent access from host supervisor 
domain components (VMM and host OS that operate in V=0, HS-mode) software.
Since in this deployment model (b), other supervisor domains have access to both 1st
and G-stage paging hardware, the root security manager (TSM-driver) must use MTT
to isolate supervisor domain memory. In this deployment model, 
TEE and TVM address spaces are identified by supervisor domain identifiers
(Smsdid) to maintain the isolation during access and in internal
address translation caches, e.g. Hart TLB lookup may be extended with the
SDID in addition to the ASID, VMID for workloads in the Confidential supervisor
domain. TVM memory isolation must support sparse memory management
models and architectural page-sizes of 4KB, 64K, 2MB, 1GB (and optionally
512GB). The hardware may implement the MTT as specified in the Smmtt 
privileged ISA extension, or other approaches may be used such as a flat 
table. The memory tracking table may be enforced at the memory controller, 
or in a page table walker.
```

information about TVM Execution and the lifecycle of a TVM is described in
the <<TVM Lifecycle>> section of this document.

=== Debug and Performance Monitoring

The following additional considerations are noted for debug and performance
monitoring:

*Debug mode considerations*

In order to support probe-mode debugging of the TSM, the RoT must support
an authorized debug of the platform. The authentication mechanism used for
debug authorization is implementation-specific, but must support the
security properties described in the Section 3.12 of the RISC-V Debug
Support specification version 1.0.0-STABLE <<R6>>. The RoT may support
multiple levels of debug authorization depending on access granted. For
probe-based debugging of the hardware, the RoT performing debug
authentication must ensure that separate attestation keys are used for TCB
reporting when probe-debug is authorized vs when the platform is not under
probe-debug mode. The probe-mode debug authorization process must invalidate
sealed keys to disallow sealed data access when in probe-debug modes. Note that
the external debug opt-in control for the hosting supervisor domain must be

> independent from the confidential supervisor domain.

Author: ```suggestion
independent from the confidential supervisor domain. Similarly, external debug 
controls should be independently managed by the RoT to allow for root security 
manager (TSM-driver) debug.
```

When a TVM is under self-hosted debugging - on a transition to TVM
execution, the TSM-driver must set up the trigger CSRs for the TVM. For TVM
debugging, the TSM-driver may inhibit M and S/HS modes in the triggers. On
transitions back to the OS/VMM, the TSM-driver will save the trigger CSRs
and associated debug states, thus not leaking any information to non-TEE
workloads. TVM self-hosted debug may be enabled from TVM creation time or
may be explicitly opted-into during execution of the TVM. The TSM may
invoke the TSM-driver to set up a TVM-specific trigger CSR state (per the
configuration of the TVM).

*Performance Monitoring considerations*

By default the TSM and all TVMs run with performance monitoring suppressed.
If a TVM runs in this default mode (opted out of performance monitoring),
on a transition to the TVM, the TSM-driver enforces this via inhibiting the
counters (using mcountinhibit).

The TVM may opt-in to use performance monitoring either at initialization or
post-initialization of the TVM.

If the TVM has opted-in to performance monitoring, the TSM may invoke the

> SBI PMU extension (via TSM-driver) or Supervisor counter delegation extension to

Author: ```suggestion
SBI PMU extension (via TSM-driver) or use Supervisor counter delegation 
(Smcdeleg) and Supervisor counter configuration (Ssccfg) extensions to
```

> SBI PMU extension (via TSM-driver) or Supervisor counter delegation extension to

Reviewer 2: Just FYI: We have assigned names for this which can be used. "Sscdeleg/Ssccfg"

Author: added

> establish a TVM-specific performance monitoring controls (counters, event
> selectors). However, the TVM must use SBI PMU extension unless TSM supports full

Author: ```suggestion
establish TVM-specific controls and configuration that allows performance 
monitoring in a TVM. However, the TVM must use SBI PMU extension unless 
TSM supports full
```

> selectors). However, the TVM must use SBI PMU extension unless TSM supports full

Reviewer 2: Do we need specify this "counters/event selectors" ? We need more state to be managed. How about  
"establish a TVM-specific state that allows performance monitoring in a TVM" or something similar

Author: added

=== TSM operation and properties

> The TSM implements security routines that are invoked by the OS/VMM or by

Author: ```suggestion
The TSM implements COVH APIs that are invoked by the OS/VMM or by
```

> the TVMs, e.g. by the VMM to grant a TVM a TEE-capable memory page and

Reviewer: security routines -> Cove TH API functions?

Author: fixed

> the TVMs, e.g. by the VMM to grant a TVM a TEE-capable memory page and

Author: ```suggestion
the TVMs, e.g. by the VMM to grant a TVM a confidential memory page and
```

setup second-stage mapping, activate a TVM virtual hart on a physical hart
etc. The TSM security routines are invoked by the OS/VMM via an ECALL with
the service call specified via registers. These service calls trap to the
TSM-driver. The TSM-driver switches hart state to the TSM context by
loading the hart's TSM execution state from the THCS.tssa and then returns
via an MRET to the TSM. The TSM executes the security routine requested
(where the TSM enforces the security properties) and may either return to
the OS/VMM via an ECALL to the TSM-driver (TEERET with reason), or may use
an SRET to return/enter into a TVM. On a subsequent TVM synchronous or
asynchronous trap (due to ECALLs or any exception/interrupt) from a TVM,

> the TSM handles the cases delegated to it by the TSM-driver (via mideleg).

Reviewer: and medeleg?

Author: yup - fixed

> the TSM handles the cases delegated to it by the TSM-driver (via mideleg).

Author: ```suggestion
the TSM handles the cases delegated to it by the TSM-driver (via mideleg and medeleg).
```

required. The TSM-driver restores the context for the OS/VMM via the
per-hart control sub-structure THCS.hssa (See <<appendix_a>>).This canonical
flow is shown in figure 3.

Beyond the basic operation described above, the following different
operational models of the TSM may be supported by an implementation:

* *Uninterruptible* *TSM* - In this model, the TSM security routines are
executed in an uninterruptible manner for S-mode interrupts (M-mode
interrupts are not inhibited). This implies that the TSM execution always
starts from a fixed initial state of the TSM harts and completes the
execution with either a TEERET to return control to the OS/VMM or via an
SRET to enter into a TVM (where the execution may be interruptible again).

* *Interruptible TSM with no re-entrancy* - In this model, after the
initial entry to the TSM with S-mode interrupts disabled, the TSM enables
interrupts during execution of the TSM security routines. The TSM may
install its interrupt handlers at this entry (or may be installed via the
TEECALL flow as shown below). On an S-mode interrupt, the TSM hart context
is saved by the TSM and keeps the interrupt pending. The TSM may then

> TEERET to the host OS/VMM with explicit information about the interruption

Reviewer: unless the interrupt is to be handled by the TSM itself?

Author: don't think we want to require any cases where the TSM handles the interrupts unilaterally.

Reviewer: How about RAS and Timer interrupts?

Author: TSM can take an action and forward the event to the host VMM

provided via the pending interrupt to the OS/VMM. The TSM-driver supports a
TEERESUME ECALL which enables the TSM to enforce that the resumption of the
interrupted TSM security routine is initiated by the OS/VMM on the same
hart. The TSM hart context restore is enforced by the TSM to allow for the
resumed TSM security routine operation to complete. An example of an
interruptible flow is the conversion of a large 2MB page to confidential

> memory, which may require a long latency encryption operation. Intermediate

Reviewer: Large memory region?

Author: yeah that example does not make sense - removing it.

> resumed TSM security routine operation to complete. An example of an
> interruptible flow is the conversion of a large 2MB page to confidential
> memory, which may require a long latency encryption operation. Intermediate

Author: ```suggestion
resumed TSM security routine operation to complete. Intermediate
```

state of the operation must be saved and restored by the TSM for such
flows.

> 

Reviewer: What software managed encryption is involved?

Author: removed that example.

  ** _If it was preemptable but not-reentrant that the next invocation on
that hart is a TEERESUME with identical parameters as the interrupted
security routine._
  ** _If the TSM was preemptable and re-entrant then the TSM would accept
both TEERESUME and TEECALL as subsequent invocations (as long as TSM
threads are available)._
* Restore the OS/VMM state saved on transition to the TSM: sstatus, stvec,
scounteren, sscratch, satp and x registers (other than a0, a1). Note that
any v/f register state must be restored by the caller.
* TSM-driver passes TSM/TVM-specified register contents to the OS/VMM to
return status from TEERET (TSM sets a0, a1 registers always - other
registers may be selected by the TVM)
* Clear Confidential mode on hart (via implementation-specific M-mode CSR to
block non-TEE mode accesses to TEE-assigned memory.)
* MRET to resume execution in OS/VMM at mepc set to THCS.hssa.pc
(THCS.hssa.pc adjusted to refer to opcode after the ECALL that triggered
the TEECALL / TEERESUME)

The TSM is stateless across TEECALL invocations, however a security routine
invoked in the TSM via a TEECALL may be interrupted and must be resumed via

> a TEERESUME i.e. _the TSM is preemptable but non-reentrant_. These

Reviewer: There is a reentrant model?
